### ISSUE TITLE
nn.Module: Confusing contract for `.register_parameter(name, None)` and `.register_buffer(name, None)`?

### ISSUE THREAD

**Issue by EricCousineau-TRI** [pytorch/pytorch#40977]
## 📚 Documentation

*Note*: I'm linking 1.5.0, not 1.5.1, b/c I can't get a permalink to 1.5.1 docs :(

I was trying to get familiar with how BatchNorm stuff works:
https://pytorch.org/docs/1.5.0/nn.html?highlight=batchnorm#torch.nn.BatchNorm1d
The docs and paper xref were great! But I didn't have a good grasp of how the running estimates were actually implemented, so I looked at the code:
https://github.com/pytorch/pytorch/blob/480851ad2c21e8e39e336d849b8030a2f91718d7/torch/nn/modules/batchnorm.py#L41-L54

Looked pretty straightforward, and I learned what "buffers" are in this context for `pytorch`; however, I was wondering what this did to the registration of buffers / parameters.

Looking at the docs for both `.register_parameter` and `.register_buffer`:
https://pytorch.org/docs/1.5.0/nn.html#torch.nn.Module.register_buffer
https://pytorch.org/docs/1.5.0/nn.html#torch.nn.Module.register_parameter

I notice that there are kinda code paths for handling the `None` case, but it's not documented (at least not in this immediate area).

From what it looks like, registering something with `None` is tantamount to just setting the attribute to `None` - there's no real exposure of the parameter to users...

## Suggestion Solutions

* Document this behavior in `.register_parameter` and `.register_buffer` - if you register a buffer / parameter with `None`, it's basically just gonna be ignored
* Have some brief exposition defining the terms "parameter" and "buffer" next to each other, and mention the possible equivalence of `Parameter.requires_grad=False` to a registered buffer? AFAICT, this comparison is done kinda implicitly, e.g.:
  * <https://pytorch.org/tutorials/beginner/saving_loading_models.html#what-is-a-state-dict>
  * <https://pytorch.org/tutorials/recipes/recipes/what_is_state_dict.html> - smells like a copy of the first?

## Related

* #8104 - all the `None` stuff
* #16675 - understandable about `register_buffers` looking like a hack... but I guess it's pretty well used?
* #39670 - behavior for `jit`
* (kinda meta) <https://discuss.pytorch.org/t/what-is-the-difference-between-register-buffer-and-register-parameter-of-nn-module/32723/10> - asked my question - why buffers instead of parameters, and equivalence
* #2018 - most recent PR that mentions `register_buffer` (Sep. 2017)
* 7f4ff0e615e - introduction of the `Module.register_buffer` in Sep. 2016, I guess to help w/ impl. of `BatchNorm`

## Example

Using Torch 1.5.1:
```py
from pprint import pformat

import torch.nn as nn

def show(code):
    module = eval(code)
    print(f">>> {code}")
    print()
    print(f"state_dict:\n{pformat(module.state_dict())}\n")
    print(f"parameters:\n{pformat(list(module.named_parameters()))}\n")
    print(f"buffers:\n{pformat(list(module.named_buffers()))}\n")
    print("---")

# Everything.
show('nn.BatchNorm1d(1)')

# No parameters.
show('nn.BatchNorm1d(1, affine=False)')

# No buffers.
show('nn.BatchNorm1d(1, track_running_stats=False)')
```

Output:
```
>>> nn.BatchNorm1d(1)

state_dict:
OrderedDict([('weight', tensor([1.])),
             ('bias', tensor([0.])),
             ('running_mean', tensor([0.])),
             ('running_var', tensor([1.])),
             ('num_batches_tracked', tensor(0))])

parameters:
[('weight', Parameter containing:
tensor([1.], requires_grad=True)),
 ('bias', Parameter containing:
tensor([0.], requires_grad=True))]

buffers:
[('running_mean', tensor([0.])),
 ('running_var', tensor([1.])),
 ('num_batches_tracked', tensor(0))]

---
>>> nn.BatchNorm1d(1, affine=False)

state_dict:
OrderedDict([('running_mean', tensor([0.])),
             ('running_var', tensor([1.])),
             ('num_batches_tracked', tensor(0))])

parameters:
[]

buffers:
[('running_mean', tensor([0.])),
 ('running_var', tensor([1.])),
 ('num_batches_tracked', tensor(0))]

---
>>> nn.BatchNorm1d(1, track_running_stats=False)

state_dict:
OrderedDict([('weight', tensor([1.])), ('bias', tensor([0.]))])

parameters:
[('weight', Parameter containing:
tensor([1.], requires_grad=True)),
 ('bias', Parameter containing:
tensor([0.], requires_grad=True))]

buffers:
[]

---
```

cc @jlin27 @albanD @mruberry

**Comment by mruberry** [comment]
Hey @EricCousineau-TRI, thanks for reporting this issue!

Our documentation could probably be clearer here. Would you like to submit a PR clarifying this behavior?

**Comment by EricCousineau-TRI** [comment]
Yup, I can give it a shot!

**Comment by jbschlosser** [comment]
There is a new [Modules note](https://pytorch.org/docs/stable/notes/modules.html#module-state) introduced in 1.8.0 that explicitly compares parameters and buffers. The behavior of passing `None` to `register_parameter()` / `register_buffer()` still needs to be documented.
